python/tvm/relay/prelude.rly

```diff
@@ -18,12 +18,290 @@
+  }
+}
+
+/*
+ * A LISP-style list ADT. An empty list is represented by `Nil`, and a member
+ * `x` can be appended to the front of a list `l` via the constructor `Cons(x, l)`.
+ */
+type List[A] {
+  Cons(A, List[A]),
+  Nil,
+}
+
+/*
+ * Get the head of a list. Assume the list has at least one element.
+ */
+def @hd[A](%xs: List[A]) -> A {
+  match? (%xs) {
+    Cons(%x, _) => %x,
+  }
+}
+
+/*
+ * Get the tail of a list.
+ */
+def @tl[A](%xs: List[A]) -> List[A] {
+  match? (%xs) {
+    Cons(_, %rest) => %rest,
+  }
+}
+
+/*
+ * Get the `n`th element of a list.
+ */
+def @nth[A](%xs: List[A], %n: Tensor[(), int32]) -> A {
+  if (%n == 0) {
+    @hd(%xs)
+  } else {
+    @nth(@tl(%xs), %n - 1)
+  }
+}
+
+/*
+ * Return the length of a list.
+ */
+def @length[A](%xs: List[A]) -> Tensor[(), int32] {
+  match (%xs) {
+    Cons(_, %rest) => 1 + @length(%rest),
+    Nil => 0,
+  }
+}
+
+/*
+ * Update the `n`th element of a list and return the updated list.
+ */
+def @update[A](%xs: List[A], %n: Tensor[(), int32], %v: A) -> List[A] {
+  if (%n == 0) {
+    Cons(%v, @tl(%xs))
+  } else {
+    Cons(@hd(%xs), @update(@tl(%xs), %n - 1, %v))
+  }
 }
 
-def @compose[a, b, c](%f: fn(b) -> c, %g: fn(a) -> b) {
-    fn (%x: a) -> c {
-        %f(%g(%x))
-    }
+/*
+ * Map a function over a list's elements. That is, `map(f, xs)` returns a new
+ * list where the `i`th member is `f` applied to the `i`th member of `xs`.
+ */
+def @map[A, B](%f: fn(A) -> B, %xs: List[A]) -> List[B] {
+  match (%xs) {
+    Cons(%x, %rest) => Cons(%f(%x), @map(%f, %rest)),
+    Nil => Nil,
+  }
+}
+
+/*
+ * A left-way fold over a list.
+ *
+ * `foldl(f, z, cons(a1, cons(a2, cons(a3, cons(..., nil)))))`
+ * evaluates to `f(...f(f(f(z, a1), a2), a3)...)`.
+ */
+def @foldl[A, B](%f: fn(A, B) -> A, %acc: A, %xs: List[B]) -> A {
+  match (%xs) {
+    Cons(%x, %rest) => @foldl(%f, %f(%acc, %x), %rest),
+    Nil => %acc,
+  }
+}
+
+/*
+ * A right-way fold over a list.
+ *
+ * `foldr(f, cons(a1, cons(a2, cons(..., cons(an, nil)))), z)`
+ * evaluates to `f(a1, f(a2, f(..., f(an, z)))...)`.
+ */
+def @foldr[A, B](%f: fn(A, B) -> B, %xs: List[A], %acc: B) -> B {
+  match (%xs) {
+    Cons(%x, %rest) => %f(%x, @foldr(%f, %rest, %acc)),
+    Nil => %acc,
+  }
+}
+
+/*
+ * A right-way fold over a nonempty list.
+ *
+ * `foldr1(f, cons(a1, cons(a2, cons(..., cons(an, nil)))))`
+ * evaluates to `f(a1, f(a2, f(..., f(an-1, an)))...)`
+ */
+def @foldr1[A](%f: fn(A, A) -> A, %xs: List[A]) -> A {
+  match? (%xs) {
+    Cons(%x, Nil) => %x,
+    Cons(%x, %rest) => %f(%x, @foldr1(%f, %rest)),
+  }
+}
+
+/*
+ * Computes the sum of a list of integer scalars.
+ */
+def @sum(%xs: List[Tensor[(), int32]]) {
+  let %add_f = fn(%x: Tensor[(), int32], %y: Tensor[(), int32]) -> Tensor[(), int32] {
+    %x + %y
+  };
+  @foldl(%add_f, 0, %xs)
+}
+
+/*
+ * Concatenates two lists.
+ */
+def @concat[A](%xs: List[A], %ys: List[A]) -> List[A] {
+  // TODO(weberlo): allow using `Cons` directly as the arg to foldr
```

Comment:
Are you encountered any problem with it @weberlo ? Iirc, the type-checking should probably still work for this case but the interpreter and other runtimes need to support that case

Reply:
Turned out to not be a problem.  I think I assumed it didn't work because the Python def was creating an `%updater` func.